### ISSUE TITLE
EasyMapping fails to map int from AFNetworking response

### ISSUE THREAD

**Issue by Heia** [lucasmedeirosleite/EasyMapping#140]
Let me start by saying this if my fist time testing out EasyMapping and as such this is from a test project using the default Core Data constructors provided by Xcode when making a new Project. You can also find this question up on [StackOverflow](http://stackoverflow.com/questions/38928088/easymapping-fails-to-map-int-from-afnetworking-response).

When I try to map an initial login of a user I get the following error:

    *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 'Unacceptable type of value for attribute:property = "userID"; desired type = NSNumber; given type = __NSArrayI; value = (
        12345
    ).'
The error is caused when trying to map the result of a POST using AFNetworking:

    Successfully logged in with responseObject: {
        collection =     {
            items =         (
                            {
                    accountid = 12345;
                    fname = John;
                    lname = Smith;
                }
            );
        };
    }
Using the following code:

    // Login server call and resulting Mapping
    [[MyHTTPSessionManager sharedManager] POST:@"users" parameters:parameters progress:^(NSProgress * _Nonnull uploadProgress) {
        
    } success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
        NSLog(@"Successfully logged in with responseObject: %@", responseObject);
        NSManagedObjectContext *context = [(AppDelegate*)[[UIApplication sharedApplication] delegate] managedObjectContext];

        // Start EasyMapping
        User *user = [EKManagedObjectMapper objectFromExternalRepresentation:responseObject withMapping:[User objectMapping] inManagedObjectContext:context];
    } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
    }];
User.m objectMapping fucntion

    + (EKManagedObjectMapping*)objectMapping
    {
        return [EKManagedObjectMapping mappingForEntityName:@"User" withRootPath:@"collection.items" withBlock:^(EKManagedObjectMapping * _Nonnull mapping) {
            [mapping mapPropertiesFromDictionary:@{
                                               @"accountid" : @"userID",
                                               @"fname"     : @"firstName",
                                               @"lname"     : @"lastName",
                                               }];
            [mapping setPrimaryKey:@"userID"];
        }];
    }
The Core data is setup such that it auto generates the following User+CoreDataProperties.h file:

    #import "User.h"

    NS_ASSUME_NONNULL_BEGIN

    @interface User (CoreDataProperties)

    @property (nullable, nonatomic, retain) NSNumber *userID;
    @property (nullable, nonatomic, retain) NSString *firstName;
    @property (nullable, nonatomic, retain) NSString *lastName;

    @end

    NS_ASSUME_NONNULL_END

I'm not sure if I'm missing something basic here or not... I am using the following API versions as described in my Podfile:
 
    pod 'AFNetworking', '~> 3.1'
    pod 'EasyMapping', '~> 0.18'

Let me know if I'm missing anything or if you have any questions on how I have things setup

**Comment by armandzwan** [comment]
Shouldn't it be:
[mapping setPrimaryKey:@"accountid"];

**Comment by Heia** [comment]
You are correct I had the wrong mapping function entirely (I was also trying a fork of EasyMapping - FastEasyMapping which was the function I had put instead). 

I updated the mapping function to the actual one used. I feel silly having done that...

**Comment by armandzwan** [comment]
You should also be wrapping Core Data calls in a performBlock
NSManagedObjectContext *context = [(AppDelegate*)[[UIApplication sharedApplication] delegate] managedObjectContext];

    // Start EasyMapping
[context performBlock:^{
    User *user = [EKManagedObjectMapper objectFromExternalRepresentation:responseObject withMapping:[User objectMapping] inManagedObjectContext:context];
}];

**Comment by Heia** [comment]
I tried updating it to have the performBlock wrapper as you suggested and the crash is still the same as before:

    *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: 
    'Unacceptable type of value for attribute: property = "userID"; desired type = NSNumber; 
    given type = __NSArrayI; value = (
        12345
    ).'

**Comment by Heia** [comment]
I re-read your response and noticed your extra suggestion at the bottom to use `arrayOfObjectsFromExternalRepresentation:` and that appears to have stopped the crash and resulted in a user if I explicitly pass the items array! However, I had to remove the `withRootPath:` from my `EKManangedObjectMapping` object constructor in `User.m`.

I guess my question then becomes how is the root path supposed to work? To me it would seem that if I left the root path and passed the `responseObject` itself then EasyMapping should be able to do the exact same, but instead it crashes with the following:

    *** Terminating app due to uncaught exception 'NSUnknownKeyException', reason:
    '[<__NSCFString 0x14665320> valueForUndefinedKey:]: this class is not key value coding-compliant for the key accountid.'
The crash still occurs even if I change to `withRootPath:@"items` and pass the external representation as `[(NSDictionary*)responseObject objectForKey:@"collection"]`

**Comment by armandzwan** [comment]
I think your withRootPath should also contain "collection". Don't know if "collection.items" will work. Try it!

**Comment by Heia** [comment]
I tried `withRootPath:@"collection"` using `NSArray *users = [EKManagedObjectMapper arrayOfObjectsFromExternalRepresentation:responseObject withMapping:[User objectMapping] inManagedObjectContext:context];` resulted in the same NSUnknownKeyException error above.

I then tried it with `User *user = [EKManagedObjectMapper objectFromExternalRepresentation:responseObject withMapping:[User objectMapping] inManagedObjectContext:context];` which resulted in no crash but an empty user (no values were mapped).

**Comment by armandzwan** [comment]
Try with "collection.items"
But you should really check the responceObject first. Is it really a Dictionary? Is Items really an Array?
Never trust a webservice. ;-)

**Comment by Heia** [comment]
So I started with the "collection.items" in the original issue which still doesn't seem to work. As for the response object itself, I originally used this webservice setup using RestKit and it seemed to have no issues. Now that I have switched to AFNetworking 3.0 + EasyMapping it seems to not work unless I pass the array itself.

Thanks for the help with locating that simply pass the nested object/array seems to work. I have been pulled away from my tests but hope to circle back and see if simply doing that will work.
